### PR TITLE
Add logs overview

### PR DESCRIPTION
**Pull request by tigrannajaryan**
The overview.md is the summarized version of ideas that were discussed
multiple times in meetings or in Gitter channels.

Note that the document is written as if we already have implemented
all desirable functionality, thus there are multiple "link TBD" entries
in the document. We will add the links as the implementations are done.

### CODE REVIEW

File: specification/logs/overview.md [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+* [Auto-Instrumenting Existing Logging](#auto-instrumenting-existing-logging)
+* [Trace Context in Legacy Formats](#trace-context-in-legacy-formats)
+
+## Introduction
+
+Of all telemetry signals logs have probably the biggest legacy. Most programming
+languages have built-in logging capabilities or well-known, widely used logging
+libraries.
+
+For metrics and traces OpenTelemetry took the approach of a clean-sheet design,
```
Comment (yurishkuro) [inline specification/logs/overview.md:24]:
```suggestion
For metrics and traces OpenTelemetry takes the approach of a clean-sheet design,
```

File: specification/logs/overview.md [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+specifies a new API and provides full implementations of this API in multiple
+languages.
+
+Our approach with logs was somewhat different. For OpenTelemetry to be
```
Comment (yurishkuro) [inline specification/logs/overview.md:28]:
```suggestion
Our approach with logs is somewhat different. For OpenTelemetry to be
```

File: specification/logs/overview.md [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+successful in logging space we needed to support existing legacy of logs and
```
Comment (yurishkuro) [inline specification/logs/overview.md:29]:
```suggestion
successful in logging space we need to support existing legacy of logs and
```

File: specification/logs/overview.md [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+- Better specified and sometimes customizable formats that can be parsed to
+  extract structured data (such as Apache logs or RFC5424 Syslog).
+
+- Formally structured formats (e.g. JSON files with well-defined schema or
+  Windows Event Log).
+
+The collection system needs to be able to discover most commonly used
+applications and have parsers that can convert these logs into a structured
+format. Like system and infrastructure logs application logs lack request
```
Comment (yurishkuro) [inline specification/logs/overview.md:236]:
```suggestion
format. Like system and infrastructure logs, application logs often lack request
```

File: specification/logs/overview.md [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+required.
+
+As opposed to manual efforts we have an interesting opportunity to "upgrade"
+application logs in a less laborious way by providing full or semi
+auto-instrumenting solutions that modify the logging libraries used by the
+application to automatically output the request context such as the trace id or
+span id with every log statement. The request context can be automatically
+extracted from incoming requests if standard compliant request propagation is
+used, e.g. via [W3C TraceContext](https://w3c.github.io/trace-context/). In
+addition the requests outgoing from the application may be injected with the
```
Comment (yurishkuro) [inline specification/logs/overview.md:268]:
```suggestion
addition, the requests outgoing from the application may be injected with the
```

File: specification/logs/overview.md [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+same request context data, thus resulting in context propagation through the
+application and creating an opportunity to have full request context in logs
+collected from all applications that can be instrumented in this manner.
+
+There are typically 2 ways to collect logs from these applications.
+
+The first approach, assuming the logs are written to files or to standard output
```
Comment (yurishkuro) [inline specification/logs/overview.md:275]:
```suggestion
The first approach, assuming the logs are written to files or to standard output,
```

File: specification/logs/overview.md [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+will require parsing the logs to convert them into more structured formats. This
```
Comment (yurishkuro) [inline specification/logs/overview.md:276]:
```suggestion
requires parsing the logs to convert them into more structured formats. This
```

File: specification/logs/overview.md [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+[then send the logs]((https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/fluentforwardreceiver))
+to OpenTelemetry Collector.
+
+The second approach is to modify the application so that the logs are output via
+a network protocol, e.g. via
+[OTLP](https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/logs/v1/logs.proto).
+The most convenient way to achieve this is to provide addons or extensions to
+the commonly used logging libraries. The addons implement sending over such
+network protocols, which would then typically require small, localized changes
+to the application code to change the logging target.
```
Comment (yurishkuro) [inline specification/logs/overview.md:292]: it might be useful to mention pros/cons of each approach

File: specification/logs/overview.md [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+![Application, API, SDK Diagram](img/application-api-sdk.png)
+
+## Events in Spans
+
+OpenTelemetry allows storing events in a span. From a data modeling perspective
+these events are a subset of the OpenTelemetry log data model. They are carried
+inside span data and can be recorded only if there is an associated span. This
+is one of the specialized use cases for logs and events that are called Embedded
+Logs. As opposed to this the traditional logs and events are not associated with
+any span and are called Standalone Logs.
```
Comment (yurishkuro) [inline specification/logs/overview.md:336]: What is the objective of this section? Should there be a guidance when Embedded vs. Standalone logs should be used? Otherwise it may be confusing to the user. Also, do we really need to introduce the term "embedded logs"?

File: specification/logs/overview.md [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+specifies a new API and provides full implementations of this API in multiple
+languages.
+
+Our approach with logs was somewhat different. For OpenTelemetry to be
+successful in logging space we needed to support existing legacy of logs and
+logging libraries, while offering improvements and better integration with the
+rest of observability world where possible.
+
+This is in essence the philosophy behind OpenTelemetry's logs support. We
+embrace existing logging solutions and make sure entire OpenTelemetry works
```
Comment (dastbe) [inline specification/logs/overview.md:34]: Seems like this should be either "make sure OpenTelemetry works" or "make sure all of Telemetry works"

File: specification/logs/overview.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 2 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 3 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 4 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 5 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 6 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 7 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 8 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:297]: Added, please see the second commit.

File: specification/logs/overview.md (hunk 9 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: You are right, it is a poor fit for this overview document. Deleted.
BTW, the Glossary already defines Embedded and Standalone Logs.

File: specification/logs/overview.md (hunk 10 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+- Free-form text formats with no easily automatable and reliable way to parse
+  structured data from them.
+
+- Better specified and sometimes customizable formats that can be parsed to
+  extract structured data (such as Apache logs or RFC5424 Syslog).
+
+- Formally structured formats (e.g. JSON files with well-defined schema or
+  Windows Event Log).
+
+The collection system needs to be able to discover most commonly used
```
Comment (zenmoto) [inline specification/logs/overview.md:233]: I think that bringing parsing the logs into scope isn't a great idea- that is a *massive* amount of work, and is what the whole business of several vendors is built on- it's also problematic in where we fit in the flow. If we parse, that parsing is baked in before it goes into Splunk/Sumo, etc, meaning that mistakes are difficult to undo, and limits how the record can be used later.

File: specification/logs/overview.md [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+More significant modifications to these applications can be done manually by
+their developers, such as addition of the request context to every log
+statement, however this is likely going to be vanishingly rare due to the effort
+required.
+
+As opposed to manual efforts we have an interesting opportunity to "upgrade"
+application logs in a less laborious way by providing full or semi
+auto-instrumenting solutions that modify the logging libraries used by the
+application to automatically output the request context such as the trace id or
+span id with every log statement. The request context can be automatically
```
Comment (zenmoto) [inline specification/logs/overview.md:264]: I think that it's probably worth mentioning here that many logging libraries are extensible in exactly this way for exactly this purpose, so it's not so much that we need to modify the logging libraries or perform agent-level manipulation of them, but we just need to write export modules for them (which will be greatly aided by having a library that manages the common functionality).

File: specification/logs/overview.md [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+extracted from incoming requests if standard compliant request propagation is
+used, e.g. via [W3C TraceContext](https://w3c.github.io/trace-context/). In
```
Comment (zenmoto) [inline specification/logs/overview.md:266]: At the logging level we can't really deal with the W3C trace propagation. If the propagation is already happening, though, we have access to it. The distinction is just that we will be looking for open trace spans, if the instrumentation doesn't exist elsewhere I don't think we would have access to it.

File: specification/logs/overview.md [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+addition, the requests outgoing from the application may be injected with the
+same request context data, thus resulting in context propagation through the
+application and creating an opportunity to have full request context in logs
+collected from all applications that can be instrumented in this manner.
+
+There are typically 2 ways to collect logs from these applications.
+
+The first approach, assuming the logs are written to files or to standard
+output, requires parsing the logs to convert them into more structured formats.
```
Comment (zenmoto) [inline specification/logs/overview.md:275]: Again, there are so many logging formats, with so many variations, I think that the best approach would be to transport them as they are, but augment with whatever data we can glean.

File: specification/logs/overview.md [hunk 15]
```diff
@@ -0,0 +1,459 @@
...
+agent to read the logs,
+[then send the logs]((https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/fluentforwardreceiver))
+to OpenTelemetry Collector.
+
+The benefit of this approach is that how logs are produced and where they are
+written by the application requires no or minimal changes. The downside is that
+it requires the often non-trivial parsing functionality. Parsing may also be
+not reliable if the output format is not well-defined.
+
+The second approach is to modify the application so that the logs are output via
```
Comment (zenmoto) [inline specification/logs/overview.md:290]: There's another option of logging into an OpenTelemetry-formatted file- I keep mentioning this as something like an OTEL-json- the OpenTelemetry logging structure serialized into a JSON format.

File: specification/logs/overview.md [hunk 16]
```diff
@@ -0,0 +1,459 @@
...
+each logging statement in the source code and instead require the developer to
+enable the OpenTelemetry SDK's logging support at the application startup. After
+that the SDK intercepts all emitted logs and modifies the emitting behavior as
+configured.
+
+### New First-Party Application Logs
+
+These are greenfield developments. OpenTelemetry provides recommendations and
+best practices about how to emit logs (along with traces and metrics) from these
+applications. For applicable languages and frameworks the auto-instrumentation
```
Comment (zenmoto) [inline specification/logs/overview.md:326]: Again, this doesn't need to be auto-instrumentation as installing an exporter module for a given logging library.

File: specification/logs/overview.md [hunk 17]
```diff
@@ -0,0 +1,459 @@
...
+will still be the easiest way to emit context-enriched logs. As already
+described earlier we provide extensions to some popular logging libraries
+languages to support the manual instrumentation cases. The extensions will
+support the inclusion of the request context in the logs and allow sending logs
+directly to the backend or to the Collector in OTLP format, bypassing the need
```
Comment (zenmoto) [inline specification/logs/overview.md:331]: I'm not sure what "directly to the backend" means here- perhaps "through existing logging transport flows" or similar?

File: specification/logs/overview.md [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+to have the logs represented as text files. Emitted logs are automatically
+augmented by application-specific resource context (e.g. process id, programming
+language, logging library name and version, etc). Full correlation across all
+context dimensions will be available for these logs.
+
+As noted earlier OpenTelemetry does not define a new logging API or create new
```
Comment (zenmoto) [inline specification/logs/overview.md:337]: I know that there are people who *do* want to build a user-focused logging library in OpenTelemetry, and some of them could be compelling- being able to do a `span.info("my logging statement")` could actually have performance benefits over looking up the active span, for example. Should we leave this open as a future enhancement?

File: specification/logs/overview.md [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+etc).
+
+OpenTelemetry recommends to collect logs from application logs using FluentBit
+or a similar agent,
+[then send](https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/fluentforwardreceiver)
+to OpenTelemetry Collector where they can be further processed and enriched.
+
+### Legacy First-Party Applications Logs
+
+These are applications that are created in-house. People tasked with setting up
```
Comment (zenmoto) [inline specification/logs/overview.md:248]: Most logging is around a free-form statement that's often in plain English. The metadata around that statement (logging level, etc.) can often be configured into a different format by configuring the logging library rather than by changing the logging statements. For example Log4J and Logback can both be configured through the classpath, so an administrator can configure them to use a JSON format by adding a properties file to the startup classpath of the application server, not even requiring changes to the deployment artifact.

File: specification/logs/overview.md (hunk 11 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:233]: The scope is very very limited though. The next paragraph says we delegate all of that to other log collection agents, such as Flientbit and won't do anything in OpenTelemetry codebase itself to do parsing. Maybe I need to make it more explicit?

File: specification/logs/overview.md [hunk 20]
```diff
@@ -0,0 +1,453 @@
...
+rest of observability world where possible.
+
+This is in essence the philosophy behind OpenTelemetry's logs support. We
+embrace existing logging solutions and make sure entire OpenTelemetry works
+nicely with existing logging libraries, log collection and processing solutions.
+
+## Limitations of non-OpenTelemetry Solutions
+
+Unfortunately existing logging solutions are currently weakly integrated with
+the rest of the observability. Logs typically have limited support in tracing
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Fixed.

File: specification/logs/overview.md (hunk 11 above)
Comment (zenmoto) [reply specification/logs/overview.md:233]: I think so- even with FluentBit, their parsing is somewhat limited. I read this section as a requirement that OT *does* do parsing of logs, and I really think that going that way is a mistake. If we have something that is already broken up at the point of logging we should avoid lossy serialization, but if what we have is already a string, I think we should focus on delivering that string plus metadata.

File: specification/logs/overview.md (hunk 15 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:290]: I agree, this is a good approach. I am not entirely certain how we will collect such logs. Should we still use FluentBit? In that case how do we ensure FluentBit parses and send JSON to Otel Collector in a way that results in exact mapping to Otel log data model? Or do we collect using Otel Collector? If we do that then we will have to implement file log reading, file tailing, understand log rotations, etc.
I would like to think through this properly before adding to the spec.

File: specification/logs/overview.md (hunk 12 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:264]: Good point. Added a paragraph after this one.

File: specification/logs/overview.md (hunk 17 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Reworded. Please check.

File: specification/logs/overview.md (hunk 11 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:233]: I added one more paragraph later in the doc to reiterate that Otel Collector will not be in the business of parsing and log file reading.

File: specification/logs/overview.md (hunk 14 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: Reworded to mention that file reading, tailing is necessary and parsing may be optionally needed. Please check.

File: specification/logs/overview.md (hunk 16 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:336]: Reworded to mention that manual installation or auto-instrumentation is possible.

File: specification/logs/overview.md (hunk 18 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:348]: It is a possibility, but I don't want to overcommit. This document is certainly not set in stone, we are free to modify it at this stage.

File: specification/logs/overview.md (hunk 19 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:248]: Yes, I think I captured what you describe in this sentence:
>People tasked with setting up
>log collection infrastructure sometimes are able to modify these applications to
>alter how logs are written and what information is included in the logs.

Do we need more clarifications?

File: specification/logs/overview.md (hunk 15 above)
Comment (zenmoto) [reply specification/logs/overview.md:300]: I think that we could use a combination of an output format that mirrored the protobuf format coupled with a FluentBit plugin that translated from the JSON format to the protobuf format when sending it to the collector could work well for this without having to duplicate FluentBit functionality.

File: specification/logs/overview.md (hunk 18 above)
Comment (kumoroku) [reply specification/logs/overview.md:348]: my recollection of the discussion so far has been to punt for now on an official OT Logging API in favor of integrating with existing libraries. I still think that this is the right decision. 

I do however also remember that we never specifically said we won't do a new API, just that we won't do it right now. I think it would be good to point that out here––I think it is a question that will come up over and over again. maybe we can have a paragraph on it here and then refer to it above (in the OpenTelemetry Solution)

File: specification/logs/overview.md [hunk 21]
```diff
@@ -0,0 +1,470 @@
...
+  this functionality may be migrated directly to the Collector.
+
+- Ability to send logs via common network protocols for logs, such as Syslog, or
+  vendor-specific log formats. Collector contains exporters that directly
+  implement this ability.
+
+To support infrastructure log and event collection in cloud native environments
+we have the following features:
+
+- Collecting Kubernetes infrastructure events and application logs from pods
```
Comment (kumoroku) [inline specification/logs/overview.md:387]: wondering why this is in this section vs. the previous one... doesn't FluentBit do at least some of this already (logs from pods, frankly not up to speed on events)

File: specification/logs/overview.md (hunk 13 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:266]: Correct, we can't deal with requests at the logging level. But if we can auto-instrument incoming requests, extract and record the trace context somewhere (e.g. in Thread Local Storage) that is later accessible to the logging library or logging appender then we can achieve this. I don't want to go into too much details in this overview document, we certainly need this all clarified elsewhere.

File: specification/logs/overview.md (hunk 18 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: I added a paragraph in OpenTelemetry Solution section to tell that adding a new API in the future is possible.

File: specification/logs/overview.md (hunk 21 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:391]: We already collect K8S metrics in the Collector. I think it is a better use experience if we also collect the events in the Collector, so that the user has a more uniform experience, instead of needing to collect metrics using Collector, collect events using FluentBit. However, you are right, in the meantime we are happy to use FluentBit's or other agent's help to collect this.

**Comment by tigrannajaryan** [comment]
@yurishkuro please take another look.

File: specification/logs/overview.md (hunk 15 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md:304]: >coupled with a FluentBit plugin that translated from the JSON format to the protobuf format

Do you mean a new plugin that will do this or there is an existing FluentBit plugin that can be configured to understand the JSON schema and output Protobufs?

Comment (zenmoto) [reply specification/logs/overview.md:304]: I'm meaning a new plugin- basically being able to buffer to files and use text-based channels to fluentd, but still be able to use the OpenTelemetry logging model.

File: specification/logs/overview.md (hunk 11 above)
Comment (avik-so) [reply specification/logs/overview.md:237]: Perhaps, change the wording from "applications and have parsers that convert" to "applications and have transformers that convert" or "have converters that convert" (or transformers than transform?)

File: specification/logs/overview.md [hunk 22]
```diff
@@ -0,0 +1,474 @@
...
+the commonly used logging libraries. The addons implement sending over such
+network protocols, which would then typically require small, localized changes
+to the application code to change the logging target.
+
+The application logs will be also enriched by the resource context, similarly to
+how it is done for third-party applications and so will potentially have full
+correlation information across all context dimensions.
+
+The downside of this approach is that the simplicity of having the logs in a
+local file is lost (e.g. ability to easily inspect the log file locally) and
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply specification/logs/overview.md:317]: Yes, that is a valid concern and saving logs to intermediary local files is listed as the first supported approach in this doc.

Note that even when bypassing the local files and sending logs from application via network it is still possible to have reasonably strong delivery guarantees if the receiving side persists the logs before confirming the receipt. In the future this can be implemented using a persistent queue in the Collector.

File: specification/logs/overview.md [hunk 23]
```diff
@@ -0,0 +1,478 @@
...
+
+### Infrastructure Logs
+
+These are logs generated by various infrastructure components, such as
+Kubernetes events (if you are wondering why events are discussed in the context
+of logs see Events and Logs (link TBD)). Like system logs, the
+infrastructure logs lack a request context and can be enriched by the resource
+context - information about the node, pod, container, etc.
+
+OpenTelemetry Collector needs to know how to query logs from most common
```
Comment (yurishkuro) [inline specification/logs/overview.md:218]: > Collector needs to know how to query logs

I think this (and system logs above) should be decoupled from the collector, similar to how various exporters of Prometheus metrics are not themselves part of the Prometheus backend, they are add-on components. Below we say that 3rd party app logs can be ingested via fluentbit or similar, why not the same for system/infra logs? There is a lot of variety among those.

File: specification/logs/overview.md (hunk 21 above)
Comment (yurishkuro) [reply specification/logs/overview.md:395]: I am not a fan of coupling k8s specifics to collector. There are many other infras that people run that is NOT k8s. Collector should first and foremost implement telemetry data plane, and expose customization via a control plane API, similar to how Envoy does it. Then k8s specific pieces can be implemented as another component that instructs the data plane, via control plane API, on where to scrape the data from and how to enrich it (e.g. resource labels).

Comment (tigrannajaryan) [reply specification/logs/overview.md]: I agree, this whole paragraph is unnecessary in this document. Deleted.

File: specification/logs/overview.md (hunk 23 above)
Comment (tigrannajaryan) [reply specification/logs/overview.md]: I agree. Reworded, please check.

**Comment by bogdandrutu** [comment]
Thanks @tigrannajaryan
